crates/ty_python_semantic/resources/mdtest/generics/pep695/variance.md

```diff
@@ -446,4 +398,101 @@ static_assert(not is_gradual_equivalent_to(D[Any], C[A]))
 static_assert(not is_gradual_equivalent_to(D[Any], C[B]))
 ```
 
+## Mutual Recursion
+
+This example due to Martin Huschenbett's PyCon 2025 talk,
+[Linear Time variance Inference for PEP 695][linear-time-variance-talk]
+
+```py
+from ty_extensions import is_assignable_to, is_equivalent_to, is_gradual_equivalent_to, is_subtype_of, static_assert, Unknown
+from typing import Any
+
+class A: ...
+class B(A): ...
+
+class C[X]:
+    def f(self) -> "D[X]":
+        return D()
+
+    def g(self, x: X) -> None: ...
+
+class D[Y]:
+    def h(self) -> C[Y]:
+        return C()
+```
+
+`X` is contravariant in both `C` and `D`:
+
+- `C` has two occurrences of `X`
+    - `X` occurs in the return type of `f` as `D[X]`
+        - `D` has one occurrence of `X`
+            - `X` occurs in the return type of `h` as `C[X]`
+    - `X` occurs contravariantly as a parameter in `g`
+
+We want the _greatest_ inferrable variance, so we start by assuming `X` occurs bivariantly in `C`.
+Therefore, `X` has one bivariant occurrence, and one contravariant occurrence, which resolves to
+contravariant.
+
+Then because `X` is contravariant in `C` `D` is as well.
+
+```py
+static_assert(not is_subtype_of(C[B], C[A]))
+static_assert(is_subtype_of(C[A], C[B]))
+static_assert(not is_subtype_of(C[A], C[Any]))
+static_assert(not is_subtype_of(C[B], C[Any]))
+static_assert(not is_subtype_of(C[Any], C[A]))
+static_assert(not is_subtype_of(C[Any], C[B]))
+
+static_assert(not is_subtype_of(D[B], D[A]))
+static_assert(is_subtype_of(D[A], D[B]))
+static_assert(not is_subtype_of(D[A], D[Any]))
+static_assert(not is_subtype_of(D[B], D[Any]))
+static_assert(not is_subtype_of(D[Any], D[A]))
+static_assert(not is_subtype_of(D[Any], D[B]))
+```
+
+## Type Variable Closure
+
+```py
+from ty_extensions import is_assignable_to, is_equivalent_to, is_gradual_equivalent_to, is_subtype_of, static_assert, Unknown
+from typing import Any
+
+class A: ...
+class B(A): ...
+
+class C[T]:
+    class D:
+        def f(self, value: T) -> None: ...
+
+    def f(self, value: D) -> None: ...
+```
+
+When inferring variance, we can't skip over classes just because they don't have type variables.
+This is because they can close over type variables, such as `D` in the above example.
```

Comment:
The last bullet point of the [scoping rules](https://typing.python.org/en/latest/spec/generics.html#scoping-rules-for-type-variables) section of the spec says that the typevar scope of an outer class does not cover an inner class, so it should be an error for `D` to refer to `T`. (We don't currently check this correctly yet.)

The pattern that they suggest for this example would be

```py
class C[T]:
    class D[U]:
        def f(self, value: U) -> None: ...

    def f(self, value: D[T]) -> None: ...
```

I'm not sure if this invalidates your reasoning in the next paragraph.